http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandlerAlgebra.scala

```diff
@@ -168,19 +167,31 @@ object AzurePubsubHandler {
         .take(size)
         .toList
 
-    val passwordChars: List[Char] = pickRandomChars(lowerLetters, 1) ++
-      pickRandomChars(upperLetters, 1) ++
-      pickRandomChars(numbers, 1) ++
-      pickRandomChars(specialChars, 1) ++
-      pickRandomChars(fullCharset, 12)
+    var password: String = pickRandomChars(fullCharset, passwordLength).mkString
+    // Keep generating passwords until we find one that has all of the required characters
+    // This is safer than picking from each subset and then shuffling
 
-    Random.shuffle(passwordChars).mkString
+    var isPasswordValid: Boolean =
+      password.exists(_.isLower) && password.exists(_.isUpper) && password.exists(_.isDigit) && password.exists(
+        specialChars.contains
+      )
+
+    while (isPasswordValid == false) {
+      password = pickRandomChars(fullCharset, passwordLength).mkString
+      isPasswordValid =
+        password.exists(_.isLower) && password.exists(_.isUpper) && password.exists(_.isDigit) && password.exists(
+          specialChars.contains
+        )
+
+    }
+    password
   }
 
   private[util] def getAzureVMSecurePassword(environment: String, sharedPassword: String): String =
     // Generate random password for Azure VM in production, for the other lower level envs we can used the shared password
+    // The password must be between 12 and 123 characters long. We are choosing 25 here
```

Comment:
sounds like this should be a random range rather than a hardcoded constant as far as security is concerned....

Reply:
I will let @sarahgibs comment on that one, my understanding was that the weakness was coming from how I was reshuffling the characters after selecting them, not how long the overall password was, but I am happy to randomly generate the length of the password too

Reply:
A set length can be fine if it's long enough. You're trying to avoid predictability of characters and brute forcing. Secure random helps avoid predictability, and the length prevents brute forcing.